From 3456f0050f889d828b7390d13d32355995b39c3f Mon Sep 17 00:00:00 2001
From: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
Date: Fri, 17 Jul 2026 22:12:32 +0000
Subject: [PATCH 099/129] net: airoha: report EN7523 GPON error counters

The OLT currently deactivates the ONU after O5 while the MAC reports RX
and FIFO errors. Add rate-limited snapshots of the BWmap FIFO state,
error control and GPON RX, TX, CRC and HEC counters when an enabled error
interrupt fires.

These counters distinguish downstream framing corruption from upstream
burst timing and QDMA delivery problems without changing datapath
behaviour.

Signed-off-by: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
---
 drivers/net/ethernet/airoha/airoha_xpon.c | 35 +++++++++++++++++++++--
 1 file changed, 32 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/airoha/airoha_xpon.c b/drivers/net/ethernet/airoha/airoha_xpon.c
index c55b9c65d3cd..b601ae74389d 100644
--- a/drivers/net/ethernet/airoha/airoha_xpon.c
+++ b/drivers/net/ethernet/airoha/airoha_xpon.c
@@ -385,8 +385,19 @@ static void airoha_xpon_phy_stop(struct device *dev, struct phy *phy,
 #define GPON_DBG_DLY		0x208
 #define GPON_DBG_IDLE_GEM_THLD	0x20C
 #define GPON_DBG_BWM_FILTER_CTRL	0x220
+#define GPON_DBG_BWM_SFIFO_STS	0x224
 #define GPON_DBG_GRP_0		0x228
 #define GPON_DBG_GRP_1		0x22C
+#define GPON_DBG_BWM_BFIFO_STS	0x250
+#define GPON_DBG_ERR_CTRL	0x260
+#define GPON_DBG_RX_GEM_CNT	0x300
+#define GPON_DBG_RX_CRC_ERR_CNT	0x304
+#define GPON_DBG_RX_GTC_CNT	0x308
+#define GPON_DBG_TX_GEM_CNT	0x30C
+#define GPON_DBG_TX_BST_CNT	0x310
+#define GPON_DBG_GEM_HEC_ONE_ERR_CNT	0x330
+#define GPON_DBG_GEM_HEC_TWO_ERR_CNT	0x334
+#define GPON_DBG_GEM_HEC_UC_ERR_CNT	0x338
 /* Debug / TX sync (EN7521 EqD adjustment) */
 #define GPON_DBG_TX_SYNC_OFFSET	0x35C
 /* Power management (always-on domain) */
@@ -2089,6 +2100,25 @@ static void gpon_restart_work_fn(struct work_struct *work)
  * Interrupt handler
  * -------------------------------------------------------------------- */
 
+static void gpon_dump_error_counters(struct gpon_priv *priv, u32 errors)
+{
+	dev_warn_ratelimited(priv->dev,
+			     "GPON error: irq=%#x sfifo=%#x bfifo=%#x ctrl=%#x\n",
+			     errors, gpon_read(priv, GPON_DBG_BWM_SFIFO_STS),
+			     gpon_read(priv, GPON_DBG_BWM_BFIFO_STS),
+			     gpon_read(priv, GPON_DBG_ERR_CTRL));
+	dev_warn_ratelimited(priv->dev,
+			     "GPON counters: rx=%u/%u/%u tx=%u/%u hec=%u/%u/%u\n",
+			     gpon_read(priv, GPON_DBG_RX_GEM_CNT),
+			     gpon_read(priv, GPON_DBG_RX_CRC_ERR_CNT),
+			     gpon_read(priv, GPON_DBG_RX_GTC_CNT),
+			     gpon_read(priv, GPON_DBG_TX_GEM_CNT),
+			     gpon_read(priv, GPON_DBG_TX_BST_CNT),
+			     gpon_read(priv, GPON_DBG_GEM_HEC_ONE_ERR_CNT),
+			     gpon_read(priv, GPON_DBG_GEM_HEC_TWO_ERR_CNT),
+			     gpon_read(priv, GPON_DBG_GEM_HEC_UC_ERR_CNT));
+}
+
 static void gpon_irq_work_fn(struct work_struct *work)
 {
 	struct gpon_priv *priv =
@@ -2150,9 +2180,8 @@ static void gpon_irq_work_fn(struct work_struct *work)
 		}
 
 		if (active & GPON_INT_ERROR_MASK)
-			dev_warn(priv->dev,
-				 "GPON MAC error interrupt: %#08x\n",
-				 active & (u32)GPON_INT_ERROR_MASK);
+			gpon_dump_error_counters(priv, active &
+					 (u32)GPON_INT_ERROR_MASK);
 
 		if (active & (INT_RX_ERR | INT_FIFO_ERR))
 			airoha_eth_xpon_dump_oam_rx_state(priv->gdm_dev);
-- 
2.55.0

